commit bdbda29c3e126754c3ae04ceb5c5d35d49aae01c
Author: Sina Chaichi Maleki <sina.chaichi2020@gmail.com>
Date:   Fri May 8 23:20:07 2026 +0200

    Fixed #29303 -- Avoided mutating original view in non_atomic_requests().
    
    Previously non_atomic_requests() mutated the _non_atomic_requests
    attribute of the original view function instead of returning a new wrapper.

diff --git a/django/db/transaction.py b/django/db/transaction.py
index fe6d3fed32..94f381779c 100644
--- a/django/db/transaction.py
+++ b/django/db/transaction.py
@@ -1,5 +1,6 @@
 import warnings
 from contextlib import ContextDecorator, contextmanager
+from functools import wraps
 
 from django.db import (
     DEFAULT_DB_ALIAS,
@@ -337,10 +338,16 @@ def atomic(using=None, savepoint=True, durable=False):
 
 def _non_atomic_requests(view, using):
     try:
-        view._non_atomic_requests.add(using)
+        databases = view._non_atomic_requests | {using}
     except AttributeError:
-        view._non_atomic_requests = {using}
-    return view
+        databases = {using}
+
+    @wraps(view)
+    def wrapper(*args, **kwargs):
+        return view(*args, **kwargs)
+
+    wrapper._non_atomic_requests = databases
+    return wrapper
 
 
 def non_atomic_requests(using=None):
